activerecord/lib/active_record/associations/preloader/through_association.rb

```diff
@@ -74,6 +74,17 @@ def source_preloaders
           end
 
           def middle_records
+            is_middle_records_loaded = true
+            middle_records = owners.flat_map do |owner|
+              association = owner.association(through_reflection.name)
+              unless association.loaded?
+                is_middle_records_loaded = false
+                break
+              end
+              Array.wrap(association.target)
+            end
+            return middle_records if is_middle_records_loaded
```

Comment:
I don't think you need the `is_middle_records_loaded` variable, because `break` will set `middle_records` to `nil`.

Reply:
Yes, either this or the one from my original suggestion: 

```ruby
middle_records || through_preloaders.flat_map(&:preloaded_records)
```

If `middle_records` is an array, it'll return the middle records array and ignore the statement after the `||`. If `middle_records` is `nil`, it'll execute the statement after the `||` and return its results.